File: Hill-Cipher/transposition.cpp
#include <iostream>
#include <vector>
#include <algorithm>

std::string transposeText(const std::string& text, const std::vector<int>& key) {
    int numColumns = key.size();
    int numRows = text.length() / numColumns;
    if (text.length() % numColumns != 0) {
        numRows += 1;
    }

    // Pad the text with spaces if necessary
    std::string paddedText = text;
    paddedText.append(numRows * numColumns - text.length(), ' ');

    // Create the transposition matrix
    std::vector<std::vector<char>> matrix;
    for (int i = 0; i < numRows; i++) {
        std::vector<char> row;
        for (int j = 0; j < numColumns; j++) {
            row.push_back(paddedText[i * numColumns + j]);
        }
        matrix.push_back(row);
    }

    // Rearrange the columns according to the key
    std::vector<int> sortedKey = key;
    std::sort(sortedKey.begin(), sortedKey.end());
    std::vector<std::vector<char>> transposedMatrix;
    for (int i = 0; i < numRows; i++) {
        std::vector<char> row;
        for (int j : sortedKey) {
            row.push_back(matrix[i][j]);
        }
        transposedMatrix.push_back(row);
    }

    // Flatten the transposed matrix into a single string
    std::string transposedText;
    for (const auto& row : transposedMatrix) {
        transposedText.append(row.begin(), row.end());
    }

    return transposedText;
}

File: Hill-Cipher/transposition.py
def transpose_text(text, key):
    num_columns = len(key)
    num_rows = len(text) // num_columns
    if len(text) % num_columns != 0:
        num_rows += 1

    # Pad the text with spaces if necessary
    text += ' ' * (num_rows * num_columns - len(text))

    # Create the transposition matrix
    matrix = []
    for i in range(num_rows):
        row = []
        for j in range(num_columns):
            row.append(text[i * num_columns + j])
        matrix.append(row)

    # Rearrange the columns according to the key
    sorted_key = sorted(range(num_columns), key=lambda k: key[k])
    transposed_matrix = []
    for i in range(num_rows):
        row = []
        for j in sorted_key:
            row.append(matrix[i][j])
        transposed_matrix.append(row)

    # Flatten the transposed matrix into a single string
    transposed_text = ''.join(''.join(row) for row in transposed_matrix)

    return transposed_text
